### PR TITLE
Fix delete nonexist oauth application 500 and prevent deadlock

### LINKED ISSUE #15356: Deleting Oauth application via api with nonexistent `id` causes internal server error
- Gitea version (or commit ref): Gitea version 1.13.7 built with GNU Make 4.1, go1.15.11 : bindata, sqlite, sqlite_unlock_notify
- Git version: git version 2.17.
- Operating system: Ubuntu 18.04
- Downloaded gitea from downloads page
- Running as a command-line program
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
- Log gist:
```
Completed DELETE /api/v1/user/applications/oauth2/x 500 Internal Server Error in 35.488257ms
```

## Description
Make a DELETE request on `/user/applications/oauth2/{id}` endpoint with supplied authorization token and invalid id in path. Then 500 response is returned with following response body:
```
curl -X DELETE 'http://localhost:2999/api/v1/user/applications/oauth2/x' -H "Authorization: token $TOKEN"
{"message":"","url":"http://localhost:3000/api/swagger"}
```
The url in the body has a different port too.

### PR DESCRIPTION
**Pull request by lunny**
Return 404 when asked to delete OAuth2 Application and ensure that the session is closed too.

Fix #15356

### CODE REVIEW

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by zeripath** [comment]
Somehow this appears to be killing CI.

**Comment by lunny** [comment]
I guess MakeRequest could not be invoke twice because req has been consumed.

File: integrations/api_oauth2_apps_test.go [hunk 1]
```diff
@@ -92,6 +92,10 @@ func testAPIDeleteOAuth2Application(t *testing.T) {
 	session.MakeRequest(t, req, http.StatusNoContent)
 
 	models.AssertNotExistsBean(t, &models.OAuth2Application{UID: oldApp.UID, Name: oldApp.Name})
+
+	req = NewRequest(t, "DELETE", urlStr)
```
Comment (6543) [inline integrations/api_oauth2_apps_test.go]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

**Review by zeripath — APPROVED** [review]

**Comment by techknowlogick** [comment]
@lunny please send backports :)

**Pull request merged**